driver/src/orderbook/streamed/state.rs

```diff
@@ -0,0 +1,531 @@
+        let balance = self.balances.entry((event.user, event.token)).or_default();
+        balance.pending_withdraw = Some(Flux {
+            amount: event.amount,
+            batch_id: event.batch_id,
+        });
+        Ok(())
+    }
+
+    fn withdraw(&mut self, event: &Withdraw, block_batch_id: BatchId) -> Result<(), Error> {
+        let balance = self.balances.entry((event.user, event.token)).or_default();
+        balance.withdraw(event.amount, block_batch_id)
+    }
+
+    fn token_listing(&mut self, event: &TokenListing) -> Result<(), Error> {
+        self.tokens.0.push((event.id, event.token));
+        Ok(())
+    }
+
+    fn order_placement(&mut self, event: &OrderPlacement) -> Result<(), Error> {
+        let order = Order {
+            buy_token: event.buy_token,
+            sell_token: event.sell_token,
+            valid_from: event.valid_from,
+            valid_until: event.valid_until,
+            price_numerator: event.price_numerator,
+            price_denominator: event.price_denominator,
+            used_amount: 0,
+        };
+        match self.orders.entry((event.owner, event.index)) {
+            Entry::Vacant(entry) => entry.insert(order),
+            Entry::Occupied(_) => return Err(Error::OrderAlreadyExists),
+        };
+        Ok(())
+    }
+
+    fn order_cancellation(
+        &mut self,
+        event: &OrderCancellation,
+        block_batch_id: BatchId,
+    ) -> Result<(), Error> {
+        let order = self
+            .orders
+            .get_mut(&(event.owner, event.id))
+            .ok_or(Error::UnknownOrder)?;
+        order.valid_until = block_batch_id - 1;
+        Ok(())
+    }
+
+    fn order_deletion(&mut self, event: &OrderDeletion) -> Result<(), Error> {
+        // Orders are allowed to be deleted multiple times.
+        self.orders.remove(&(event.owner, event.id));
+        Ok(())
+    }
+}
+
+#[derive(Clone, Copy, Debug, Error)]
+pub enum Error {
+    #[error("unknown token")]
+    UnknownToken,
+    #[error("unknown order")]
+    UnknownOrder,
+    #[error("order already exists")]
+    OrderAlreadyExists,
+    #[error("math underflow")]
+    MathUnderflow,
+}
+
+/// Bidirectional map between token id and token address.
+///
+/// Std does not have this type so we use a vector. Alternatively we could find a crate.
+#[derive(Clone, Debug, Default, Deserialize, Serialize)]
+struct Tokens(Vec<(TokenId, TokenAddress)>);
+
+impl Tokens {
+    fn get_address_by_id(&self, id: TokenId) -> Option<Address> {
+        self.0
+            .iter()
+            .find(|(id_, _)| *id_ == id)
+            .map(|(_, address)| *address)
+    }
+
+    fn get_id_by_address(&self, address: TokenAddress) -> Option<TokenId> {
+        self.0
+            .iter()
+            .find(|(_, address_)| *address_ == address)
+            .map(|(token_id, _)| *token_id)
+    }
+}
+
+#[derive(Copy, Clone, Debug, Deserialize, Serialize)]
+struct Order {
```

Comment:
nit: This is a data type, and I think it might make sense for it to live with the other type aliases `TokenId` and `BatchId`. Maybe `Flux` and `Balance` as well? That way this module focuses more on the logic around keeping the data in sync with the EVM.

Reply:
It is specific to this file (not pub) so I want to keep it here.

Reply:
Ah, good point. never mind.